sda-commons-client-jersey/src/main/java/org/sdase/commons/client/jersey/filter/TraceTokenClientFilter.java

```diff
@@ -18,42 +13,22 @@
  */
 public class TraceTokenClientFilter implements ClientRequestFilter, ClientResponseFilter {
 
-  private static final String ADDED_TO_MDC_FOR_REQUEST =
-      TraceTokenClientFilter.class.getName() + "_ADDED_TO_MDC";
+  private static final String TRACE_TOKEN_CONTEXT =
+      TraceTokenClientFilter.class.getName() + "_TRACE_TOKEN_CONTEXT";
 
   @Override
   public void filter(ClientRequestContext requestContext) {
-    String traceToken =
-        takeTraceTokenFromIncomingRequest()
-            .orElseGet(() -> createTraceTokenForOutgoingRequest(requestContext));
-    requestContext.getHeaders().add(RequestTracing.TOKEN_HEADER, traceToken);
+    var traceTokenContext = TraceTokenContext.getOrCreateTraceTokenContext();
+    String traceToken = traceTokenContext.get();
+    requestContext.getHeaders().add(TraceTokenContext.TRACE_TOKEN_HTTP_HEADER_NAME, traceToken);
+    requestContext.setProperty(TRACE_TOKEN_CONTEXT, traceTokenContext);
   }
 
   @Override
   public void filter(ClientRequestContext requestContext, ClientResponseContext responseContext) {
-    if (Boolean.TRUE.equals(requestContext.getProperty(ADDED_TO_MDC_FOR_REQUEST))) {
-      requestContext.removeProperty(ADDED_TO_MDC_FOR_REQUEST);
-      MDC.getMDCAdapter().remove(RequestTracing.TOKEN_MDC_KEY);
-    }
-  }
-
-  private String createTraceTokenForOutgoingRequest(ClientRequestContext clientRequestContext) {
-    String traceToken = UUID.randomUUID().toString();
-    addNewTokenToMdcForCurrentRequest(clientRequestContext, traceToken);
-    return traceToken;
-  }
-
-  private void addNewTokenToMdcForCurrentRequest(
-      ClientRequestContext requestContext, String newTraceToken) {
-    if (MDC.getMDCAdapter() != null) {
-      MDC.put(RequestTracing.TOKEN_MDC_KEY, newTraceToken);
-      requestContext.setProperty(ADDED_TO_MDC_FOR_REQUEST, Boolean.TRUE);
+    if (requestContext.getProperty(TRACE_TOKEN_CONTEXT)
+        instanceof TraceTokenContext traceTokenContext) {
+      traceTokenContext.closeIfCreated();
```

Comment:
What if the client does not respond, is then the trace context is not cleared?
If that all in a separate thread i guess that would not be a problem

Reply:
> What if the client does not respond, is then the trace context is not cleared?

You are right. I did not expect that this kind of filter is only called when the server really responds. But that actually happened in the previous implementation as well.

> If that all in a separate thread i guess that would not be a problem

There could be a problem with a long lasting trace token in special cases. But I would assume that in most cases a thread will be terminated after a single action when it is not triggered with a trace token context that is around this request.

However, a quick research shows, that you can't cover the error case of timeout or connection issues. We could only change the behavior and avoid creating a trace token if we are not already in a context with a request token.

What do you think is the better option?

Reply:
Good idea. I added [this comment](https://github.com/SDA-SE/sda-dropwizard-commons/pull/3243/commits/2e224c3345b975dffdf31b6fc8ec81d9cca1e894).